onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -511,12 +520,17 @@ class PlannerImpl {
 
       auto outputs = pnode->OutputDefs();
       auto num_outputs = outputs.size();
+      bool external_outputs = ExternalOutputs(*pnode);
       for (size_t i = 0; i < num_outputs; ++i) {
         auto* node_output = outputs[i];
         if (!node_output->Exists()) continue;
         OrtValueIndex index = Index(node_output->Name());
         ProcessDef(index, node_output);
         ++UseCount(index);
+        if (external_outputs) {
+          // Ensures external outputs will not be reused.
+          ++UseCount(index);
```

Comment:
with the shape inferencing logic fixed in Yield's schema. Do we still need this fix? 
IIUC, this additional use count is blocking the memory planner from use this buffer. But if our previous analysis is right on "output_grads couldn't have other consumer other than the backward() function", do this mean, in theory, we can still reuse the output_grads' buffer in the backward graph?

Reply:
but this is indeed a dangerous operation, as we are stealing to use memory allocated by pytorch.

Reply:
I think we still need this if we don't want to change/overwrite the tensors from outside (just like inputs). Not sure if "output_grads couldn't have other consumer other than the backward() function" is always true, maybe OK for this PyTorch case, but logically we should not change external tensors...

Reply:
@SherlockNoMad  This is correct and needed for PreExisting tensors, please see input tensors are incremented twice because they are initialized to count of 1 [here](https://github.com/microsoft/onnxruntime/blob/67c67408c4328adcd5b6e14938516fa4e19c82a6/onnxruntime/core/framework/allocation_planner.cc#L437) and then whenever any node consumes them the count is again incremented [here](https://github.com/microsoft/onnxruntime/blob/67c67408c4328adcd5b6e14938516fa4e19c82a6/onnxruntime/core/framework/allocation_planner.cc#L507), this ensures the count never drops to zero so they are never freeded or reused. PreExisting by definition means we should not be freeing or reusing tensor buffer because the ownership is with the user. Now with these PyTorch tensors who owns them? If you think ORT owns them then you are correct we should not increment twice because that will block the tensor buffer from being reused but your comment seems to imply otherwise.

@iK1D I would clean the code a little and just increment in one shot, i.e `if (external_outputs) { UseCount(index) += 2; } else {++UseCount(index);}` and please add a test :)

Reply:
Upon on further thinking, I feel we should try and own these tensors so we can reuse these tensors in backward pass just like we did without ORTModule (i.e loss op's outputs were reused) otherwise we could a potentially lose on max batch size. However if we went that route we will need to introduce a new type of memory that indicates its an preexisting tensor whose ownership is with ORT, i.e something like PreExistingORT

Reply:
I agree with both of your thinkings. 
If we treat this as external tensor, the use_count should be 2, but this will limit ORT from reusing this buffer. 
If we wish to treat this as ORT owned tensor, we need a way to "std::move()" the output_grads ownership into ORT, but so far we haven't find the way. 

Let's keep this problem in the back of our mind, and see we can come up with something smart.